### PR TITLE
Eps32 Lilygo t5v2 BSP

### PR DESCRIPTION
**Pull request by adamkaliszan**
## Summary
BSP for Lilygo T5v2 board (e-paper display with SD card reader).
## Impact
None
## Testing
Yes, on Lilygo T5v2 board.

### CODE REVIEW

File: boards/xtensa/esp32/common/include/esp32_ssd1680.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __BOARDS_ARM_STM32_COMMON_INCLUDE_STM32_SSD1680_H
+#define __BOARDS_ARM_STM32_COMMON_INCLUDE_STM32_SSD1680_H
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/common/include/esp32_ssd1680.h:22]:
```suggestion
#ifndef __BOARDS_XTENSA_ESP32_COMMON_INCLUDE_ESP32_SSD1680_H
#define __BOARDS_XTENSA_ESP32_COMMON_INCLUDE_ESP32_SSD1680_H
```

File: boards/xtensa/esp32/common/include/esp32_ssd1680.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ ****************************************************************************/
+
+FAR struct lcd_dev_s *board_ssd1680_getdev(void);
+
+#undef EXTERN
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* __BOARDS_ARM_STM32_COMMON_INCLUDE_STM32_SSD1680_H */
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/common/include/esp32_ssd1680.h:64]:
```suggestion
#endif /* __BOARDS_XTENSA_ESP32_COMMON_INCLUDE_ESP32_SSD1680_H */
```

File: boards/xtensa/esp32/common/include/esp32_ssd1680.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * Description:
+ *   Get the SSD1680 device driver instance
+ *
+ * Returned Value:
+ *   Pointer to the instance
+ *
+ ****************************************************************************/
+
+FAR struct lcd_dev_s *board_ssd1680_getdev(void);
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/common/include/esp32_ssd1680.h:57]:
```suggestion
struct lcd_dev_s *board_ssd1680_getdev(void);
```

File: boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_boot.c [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+#ifdef CONFIG_SMP
+  /* To avoid corrupting the heap, this region of memory (~3KB) is not
+   * included until the APP CPU has started.
+   * So we can't add it with the rest of the regions at xtensa_add_region(),
+   * that function is called early from up_initialize().  We wait until the
+   * SMP bringup is complete.
+   */
+
+  umm_addregion((FAR void *)HEAP_REGION_ROMAPP_START,
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_boot.c:92]:
```suggestion
  umm_addregion((void *)HEAP_REGION_ROMAPP_START,
```

File: boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_buttons.c [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+ *   board_button_irq() may be called to register an interrupt handler that
+ *   will be called when a button is depressed or released.  The ID value is
+ *   a button enumeration value that uniquely identifies a button resource.
+ *   See the BUTTON_* definitions in board.h for the meaning of enumeration
+ *   value.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_IRQBUTTONS
+int board_button_irq(int id, xcpt_t irqhandler, FAR void *arg)
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_buttons.c:132]:
```suggestion
int board_button_irq(int id, xcpt_t irqhandler, void *arg)
```

File: boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_gpio.c [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+  struct esp32gpio_dev_s esp32gpio;
+  pin_interrupt_t callback;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+#if BOARD_NGPIOOUT > 0
+static int gpout_read(FAR struct gpio_dev_s *dev, FAR bool *value);
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_gpio.c:90]: ramove `FAR` here and in other places in this file

File: boards/xtensa/esp32/common/include/esp32_ssd1680.h (hunk 1 above)
Comment (adamkaliszan) [reply boards/xtensa/esp32/common/include/esp32_ssd1680.h]: Thx. shame on me... It is fixed now.

File: boards/xtensa/esp32/ttgo_eink5_v2/scripts/Make.defs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	LDSCRIPT_TEMPLATE = $(BOARD_DIR)$(DELIM)scripts$(DELIM)esp32.template.ld
+else
+	LDSCRIPT_TEMPLATE = $(BOARD_COMMON_DIR)$(DELIM)scripts$(DELIM)esp32.template.ld
+endif
+
+ifeq ($(CONFIG_DEBUG_SYMBOLS),y)
+  ARCHOPTIMIZATION = -g
+endif
+
+ifeq ($(CONFIG_STACK_CANARIES),y)
```
Comment (xiaoxiang781216) [inline boards/xtensa/esp32/ttgo_eink5_v2/scripts/Make.defs:60]: many setting is moved to Toolchain.defs, so please update Make.defs by referencing other similar files on the master

**Review by xiaoxiang781216 — APPROVED** [review]

File: boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_bringup.c [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+#endif
+
+#if defined(CONFIG_ESP32_EFUSE)
+#include <nuttx/efuse/efuse.h>
+#endif
+#include <nuttx/fs/fs.h>
+#include <nuttx/himem/himem.h>
+
+#if defined(CONFIG_ESP32_EFUSE)
+#include "esp32_efuse.h"
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_bringup.c:51]:
```suggestion
#  include "esp32_efuse.h"
```

File: boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_bringup.c [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+#include <stdio.h>
+
+#include <errno.h>
+
+#if defined(CONFIG_VIDEO_FB) && defined(CONFIG_LCD_FRAMEBUFFER)
+#include <nuttx/video/fb.h>
+#endif
+
+#if defined(CONFIG_ESP32_EFUSE)
+#include <nuttx/efuse/efuse.h>
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/ttgo_eink5_v2/src/esp32_bringup.c:45]: can be merged to `#if defined(CONFIG_ESP32_EFUSE)` below

**Review by pkarashchenko — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

File: boards/xtensa/esp32/common/src/esp32_ssd1680.c [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+  if (!spi)
+    {
+      lcderr("ERROR: Failed to initialize SPI port %d\n",
+        CONFIG_SSD1680_SPI_BUS);
+      return -ENODEV;
+    }
+  else
+    {
+      lcdinfo("Using SPI bus %d. SPI is initialized\n",
+        CONFIG_SSD1680_SPI_BUS);
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/common/src/esp32_ssd1680.c:147]:
```suggestion
      lcdinfo("Using SPI bus %d. SPI is initialized\n",
              CONFIG_SSD1680_SPI_BUS);
```

File: boards/xtensa/esp32/common/src/esp32_ssd1680.c [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+  lcdinfo("Read busy line is disabled\n");
+#endif
+
+  /* Initialize SPI */
+
+  spi = esp32_spibus_initialize(CONFIG_SSD1680_SPI_BUS);
+  if (!spi)
+    {
+      lcderr("ERROR: Failed to initialize SPI port %d\n",
+        CONFIG_SSD1680_SPI_BUS);
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/common/src/esp32_ssd1680.c:141]:
```suggestion
      lcderr("ERROR: Failed to initialize SPI port %d\n",
             CONFIG_SSD1680_SPI_BUS);
```

File: boards/xtensa/esp32/common/src/esp32_ssd1680.c [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+  /* Bind the SPI port to the E-PAPER display */
+
+  g_lcddev = ssd1680_initialize(spi, &g_ssd1680_priv);
+  if (!g_lcddev)
+    {
+      lcderr("ERROR: Failed to bind SPI port %d to E-paper display\n",
+          CONFIG_SSD1680_SPI_BUS);
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/common/src/esp32_ssd1680.c:156]:
```suggestion
      lcderr("ERROR: Failed to bind SPI port %d to E-paper display\n",
             CONFIG_SSD1680_SPI_BUS);
```

File: boards/xtensa/esp32/ttgo_eink5_v2/src/ttgo_eink5_v2.h [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __BOARDS_XTENSA_ESP32_TTGO_EINK5_V2_SRC_TTGO_EINK5_V2_SRC_H
+#define __BOARDS_XTENSA_ESP32_TTGO_EINK5_V2_SRC_TTGO_EINK5_V2_SRC_H
```
Comment (pkarashchenko) [inline boards/xtensa/esp32/ttgo_eink5_v2/src/ttgo_eink5_v2.h:22]:
```suggestion
#ifndef __BOARDS_XTENSA_ESP32_TTGO_EINK5_V2_SRC_TTGO_EINK5_V2_H
#define __BOARDS_XTENSA_ESP32_TTGO_EINK5_V2_SRC_TTGO_EINK5_V2_H
```

**Review by pkarashchenko — COMMENTED** [review]
In general looks good. Only few minor comments

File: boards/xtensa/esp32/common/src/esp32_ssd1680.c (hunk 10 above)
Comment (adamkaliszan) [reply boards/xtensa/esp32/common/src/esp32_ssd1680.c]: OK. I have applied suggestions. Thx. for Your remark.

File: boards/xtensa/esp32/common/src/esp32_ssd1680.c (hunk 11 above)
Comment (adamkaliszan) [reply boards/xtensa/esp32/common/src/esp32_ssd1680.c]: OK. I have applied suggestions. Thx. for Your remark.

File: boards/xtensa/esp32/common/src/esp32_ssd1680.c [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+  lcdinfo("Using pin %d as RESET\n", CONFIG_SSD1680_GPIO_PIN_RST);
+#elif
+  lcdinfo("RESET line is disabled\n");
+#endif
+
+#if defined(CONFIG_SSD1680_GPIO_PIN_BUSY) && \
+  (CONFIG_SSD1680_GPIO_PIN_BUSY >= 0)
+  esp32_configgpio(CONFIG_SSD1680_GPIO_PIN_BUSY, INPUT | PULLUP);
+  lcdinfo("Using pin %d for reading busy state\n",
+    CONFIG_SSD1680_GPIO_PIN_BUSY);
```
[the remark this answers is not in the document]
Comment (adamkaliszan) [reply boards/xtensa/esp32/common/src/esp32_ssd1680.c]: OK. I have applied suggestions. Thx. for Your remark.

File: boards/xtensa/esp32/ttgo_eink5_v2/src/ttgo_eink5_v2.h [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+ *   to indicate the nature of any failure.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SPI_DRIVER
+int board_spidev_initialize(int bus);
+#endif
+
+#endif /* __ASSEMBLY__ */
+#endif /* __BOARDS_XTENSA_ESP32_TTGO_EINK5_V2_SRC_TTGO_EINK5_V2_SRC_H*/
```
[the remark this answers is not in the document]
Comment (adamkaliszan) [reply boards/xtensa/esp32/ttgo_eink5_v2/src/ttgo_eink5_v2.h]: OK. I have applied suggestions. Thx. for Your remark.

File: boards/xtensa/esp32/ttgo_eink5_v2/src/ttgo_eink5_v2.h (hunk 13 above)
Comment (adamkaliszan) [reply boards/xtensa/esp32/ttgo_eink5_v2/src/ttgo_eink5_v2.h]: OK. I have applied suggestions. Thx. for Your remark.

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**